src/compute-types/src/dataflows.rs

```diff
@@ -439,26 +439,90 @@ where
     }
 }
 
-impl<P: PartialEq, S: PartialEq, T: timely::PartialOrder> DataflowDescription<P, S, T> {
+impl<P, S, T> DataflowDescription<P, S, T>
+where
+    P: PartialEq,
+    S: Clone + PartialEq,
+    T: Clone + timely::PartialOrder,
+{
     /// Determine if a dataflow description is compatible with this dataflow description.
     ///
     /// Compatible dataflows have equal exports, imports, and objects to build. The `as_of` of
     /// the receiver has to be less equal the `other` `as_of`.
     ///
     // TODO: The semantics of this function are only useful for command reconciliation at the moment.
     pub fn compatible_with(&self, other: &Self) -> bool {
-        let equality = self.index_exports == other.index_exports
-            && self.sink_exports == other.sink_exports
-            && self.objects_to_build == other.objects_to_build
-            && self.index_imports == other.index_imports
-            && self.source_imports == other.source_imports;
-        let partial = if let (Some(as_of), Some(other_as_of)) = (&self.as_of, &other.as_of) {
-            timely::PartialOrder::less_equal(as_of, other_as_of)
+        let old = self.as_comparable();
+        let new = other.as_comparable();
+
+        let equality = old.index_exports == new.index_exports
+            && old.sink_exports == new.sink_exports
+            && old.objects_to_build == new.objects_to_build
+            && old.index_imports == new.index_imports
+            && old.source_imports == new.source_imports;
+
+        let partial = if let (Some(old_as_of), Some(new_as_of)) = (&old.as_of, &new.as_of) {
+            timely::PartialOrder::less_equal(old_as_of, new_as_of)
         } else {
             false
         };
+
         equality && partial
     }
+
+    /// Returns a `DataflowDescription` that has the same semantics as `self` and can be
+    /// semantically compared to other `DataflowDescription`s.
+    ///
+    /// For now this method performs a single normalization only: It replaces transient `GlobalId`s
+    /// that are only used internally (i.e. not imported nor exported) with consecutive IDs
+    /// starting from `t1`.
+    pub fn as_comparable(&self) -> DataflowDescription<&P, S, T> {
+        let external_ids: BTreeSet<_> = self.import_ids().chain(self.export_ids()).collect();
+
+        let mut id_counter = 0;
+        let mut replacements = BTreeMap::new();
+
+        let mut maybe_replace = |id: GlobalId| {
+            if id.is_transient() && !external_ids.contains(&id) {
+                *replacements.entry(id).or_insert_with(|| {
+                    id_counter += 1;
+                    GlobalId::Transient(id_counter)
+                })
+            } else {
+                id
+            }
+        };
+
+        let objects_to_build = self
+            .objects_to_build
+            .iter()
+            .map(|obj| BuildDesc {
+                id: maybe_replace(obj.id),
+                plan: &obj.plan,
```

Comment:
Surprised to see that we don't have to replace any ID in the plan. Is this because we only use transient IDs between dataflows and sinks, or dataflows and peeks? Otherwise, there might be some part in `obj.plan` that mentions the now-replaced ID and hence would not compare correctly in the test above in `compatible_with`.

Reply:
`Plan`s in `objects_to_build` either refer to objects imported through `index_imports` and `source_imports`, or to other objects in `objects_to_build`. The first group is excluded here through the `external_ids` check, so we only need to look at the second case.

AFAICT the only situation where `Plan`s reference other `objects_to_build` today is when the dataflow consists of multiple view and the top-level view draws from outputs of lower-level views. Views always have non-transient IDs, so these references should never use transient `GlobalId`s.

Of course this might change in the future. If it does, reconciliation will start failing sometimes, which fortunately only regresses performance, not correctness. Still, I'd also feel more comfortable with a more robust solution.

Reply:
I've pushed a fixup commit that also replaces IDs within the plans. The only drawback is that we now have to clone the plans, so this becomes a bit more expensive. But then again, we only call this method during reconciliation, so it's probably not a big deal.